[widget audit] toga.Table (#2011)
beeware/toga: docs/reference/api/widgets/table.rst

we will display a table of 2 columns, with 3 initial rows of data:

.. code-block:: python

    import toga

    table = toga.Table(
        headings=["Name", "Age"],
        data=[
            ("Arthur Dent", 42),
            ("Ford Prefect", 37),
            ("Tricia McMillan", 38),
        ]
    )

    # Get the details of the first item in the data:
    print(f"{table.data[0].name} is age {table.data[0].age}")

    # Append new data to the table
    table.data.append(("Zaphod Beeblebrox", 47))

You can also specify data for a Table using a list of dictionaries. This allows you to
store data in the data source that won't be displayed in the table. It also allows you
to control the display order of columns independent of the storage of that data.

.. code-block:: python

    import toga

    table = toga.Table(
        headings=["Name", "Age"],
        data=[
            {"name": "Arthur Dent", "age": 42, "planet": "Earth"},
            {"name", "Ford Prefect", "age": 37, "planet": "Betelgeuse Five"},
            {"name": "Tricia McMillan", "age": 38, "plaent": "Earth"},
        ]
    )

    # Get the details of the first item in the data:

>     print(f"{table.data[0].name}, who is age {table.data[0].age}, "
>           f"is from {table.data[0].planet}")

Author: Our example code should be black-compatible:
```suggestion
    print(
        f"{table.data[0].name}, who is age {table.data[0].age}, "
        f"is from {table.data[0].planet}"
    )
```

Reviewer: Even better, I've factored out `table.data[0]` into a variable so it doesn't even need to wrap anymore.

The attribute names used on each row of data (called "accessors") are created
automatically from the headings that you provide. If you want to use different
attributes, you can override them by providing an ``accessors`` argument. In this
example, the table will use "Name" as the visible header, but internally, the attribute
"character" will be used:

.. code-block:: python

    import toga

    table = toga.Table(
        headings=["Name", "Age"],
        accessors={"Name", 'character'},
        data=[
            {"character": "Arthur Dent", "age": 42, "planet": "Earth"},
            {"character", "Ford Prefect", "age": 37, "planet": "Betelgeuse Five"},
            {"name": "Tricia McMillan", "age": 38, "plaent": "Earth"},
        ]
    )

    # Get the details of the first item in the data:
    print(f"{table.data[0].character}, who is age {table.data[0].age}, "
          f"is from {table.data[0].planet}")

The value provided by an accessor is interpreted as follows:

> * If the value is a :any:`Widget`, that widget will be displayed in the cell. Note that
>   this is currently a beta API, is currently only supported on macOS, and may change in
>   future.

Author: This was previously in a notes section; I agree it's worth flagging as a beta feature here, I think having the separate notes block that explains exactly what "beta" means in this context is worthwhile.
```suggestion
* If the value is a :any:`Widget`, that widget will be displayed in the cell. Note that
  this is currently a beta API; see the :ref:`Notes section <table-notes>` for details.
```

(with a notes section describing macOS limitations, etc)

Reviewer: I know we normally put OS-specific notes in a separate section, but this is a feature which quite possibly will never be implemented on any other platform, so I think that should be stated up front.

Author: I agree that it should be stated up front that this is beta, and will almost certainly undergo a radical change before becoming final. What I'm suggesting is that we make that a general comment about API stability in the "main" docs, and leave the platform-specificity to the notes section. However, even if we do mention the macOS specificity here, we've got a general pattern of "platform-specific weirdness is in the notes section"; so I think we should have a Notes section highlighting the macOS restriction.

Reviewer: OK, done.

    )

    # Get the details of the first item in the data:
    print(f"{table.data[0].name}, who is age {table.data[0].age}, is from {table.data[0].planet}")
    print(f"{table.data[0].name}, who is age {table.data[0].age}, "
          f"is from {table.data[0].planet}")

The attribute names used on each row of data (called "accessors") are created automatically from

> the name of the headings that you provide. This is done by:

Reviewer: I moved some details into the constructor docstrings to make them self-contained, leaving the Usage section as mostly just examples.
